src/server/stream_family.cc

```diff
@@ -1192,97 +1242,255 @@ void StreamFamily::XRevRange(CmdArgList args, ConnectionContext* cntx) {
-          kSyntaxErr);
+      // Set ID to 0 so if the ID cannot be resolved (when the stream doesn't
+      // exist) it takes the first entry added.
+      id.val.ms = 0;
+      id.val.seq = 0;
+      id.last_id = true;
+      opts.stream_ids.emplace(key, id);
+      continue;
     }
 
     if (idstr == ">") {
       // XREADGROUP is not supported.
-      return (*cntx)->SendError(
+      (*cntx)->SendError(
           "The > ID can be specified only when calling XREADGROUP using the GROUP <group> "
           "<consumer> option.",
           kSyntaxErr);
+      return std::nullopt;
     }
 
-    ParsedStreamId id;
     if (!ParseID(idstr, true, 0, &id)) {
-      return (*cntx)->SendError(kInvalidStreamId, kSyntaxErrType);
+      (*cntx)->SendError(kInvalidStreamId, kSyntaxErrType);
+      return std::nullopt;
     }
+
     // We only include messages with IDs greater than start so increment the
     // starting ID.
     streamIncrID(&id.val);
-    read_opts.stream_ids.emplace(key, id);
+    opts.stream_ids.emplace(key, id);
   }
 
-  unsigned shard_count = shard_set->size();
-  vector<vector<RecordVec>> xread_resp(shard_count);
+  return opts;
+}
+
+// Returns the last ID of each stream in the transaction.
+OpResult<unordered_map<string_view, streamID>> StreamLastIDs(Transaction* trans) {
+  vector<OpResult<unordered_map<string_view, streamID>>> last_ids_res(shard_set->size());
+
   auto cb = [&](Transaction* t, EngineShard* shard) {
     ShardId sid = shard->shard_id();
-    xread_resp[sid] = OpRead(t->GetOpArgs(shard), t->GetShardArgs(shard->shard_id()), read_opts);
+    last_ids_res[sid] = OpLastIDs(t->GetOpArgs(shard), t->GetShardArgs(shard->shard_id()));
     return OpStatus::OK;
   };
-  OpStatus result = cntx->transaction->ScheduleSingleHop(std::move(cb));
-  CHECK_EQ(OpStatus::OK, result);
+  trans->Execute(std::move(cb), false);
+
+  unordered_map<string_view, streamID> last_ids;
+  for (auto res : last_ids_res) {
+    if (!res) {
+      return res.status();
+    }
+
+    for (auto& e : *res) {
+      last_ids.emplace(e.first, e.second);
+    }
+  }
+  return last_ids;
+}
+
+void StreamFamily::XRead(CmdArgList args, ConnectionContext* cntx) {
+  auto opts = ParseReadArgs(args, cntx);
+  if (!opts) {
+    return;
+  }
+
+  cntx->transaction->Schedule();
+
+  auto last_ids = StreamLastIDs(cntx->transaction);
+  if (!last_ids) {
+    // Close the transaction.
+    auto close_cb = [&](Transaction* t, EngineShard* shard) { return OpStatus::OK; };
+    cntx->transaction->Execute(std::move(close_cb), true);
+
+    if (last_ids.status() == OpStatus::WRONG_TYPE) {
+      (*cntx)->SendError(kWrongTypeErr);
+      return;
+    }
+
+    return (*cntx)->SendNullArray();
+  }
+
+  // Resolve '$' IDs and check if there are any streams with entries that can
+  // be resolved without blocking.
+  bool block = true;
+  for (auto& e : opts->stream_ids) {
+    string_view stream = e.first;
+    ParsedStreamId& requested_id = e.second;
```

Comment:
Small note about style: Compact code is often easier to read and handle (not saying our existing is perfect - by far not 🙂 )

For example, here `for (auto& [stream, id] : opts->stream_ids)` does the job, likewise with manual loops or one-step intermediate variables above.

Reply:
> here for (auto& [stream, id] : opts->stream_ids) does the job

Nice thanks, I didn't know that syntax existed, updated. And fixed the loop above